File: src/Cistern.Linq/Links/Take.cs
using System;

namespace Cistern.Linq.Links
{
    sealed class Take<T>
        : ILink<T, T>
        , Optimizations.ILinkFastCount
        , Optimizations.IMergeSkipTake<T>
    {
        private int _count;

        public Take(int count) =>
            _count = count;

        bool Optimizations.ILinkFastCount.SupportedAsConsumer => true;

        int? Optimizations.ILinkFastCount.FastCountAdjustment(int count) => Math.Min(_count, count);

        public IConsumable<T> MergeSkip(Consumable<T> consumable, int skip)
        {
            if (skip == 0)
                return consumable;

            return consumable.ReplaceTailLink(new TakeSkip<T>(_count, skip));
        }

        public IConsumable<T> MergeTake(Consumable<T> consumable, int take)
        {
            if (take > _count)
                return consumable;

            return consumable.ReplaceTailLink(new Take<T>(take));
        }

        Chain<T> ILink<T,T>.Compose(Chain<T> activity) =>
            new Activity(_count, activity);

        sealed class Activity
            : Activity<T, T>
            , Optimizations.IHeadStart<T>
        {
            private readonly int count;

            private int index;

            public Activity(int count, Chain<T> next) : base(next) =>
                (this.count, index) = (count, 0);

            public override ChainStatus ProcessNext(T input)
            {
                if (index >= count)
                {
                    return ChainStatus.Stop;
                }

                checked
                {
                    index++;
                }

                if (index >= count)
                    return ChainStatus.Stop | Next(input);
                else
                    return Next(input);
            }

            ChainStatus Optimizations.IHeadStart<T>.Execute(ReadOnlySpan<T> source)
            {
                if (count < source.Length)
                    source = source.Slice(0, count);

                if (next is Optimizations.IHeadStart<T> optimizations)
                    return optimizations.Execute(source);
                else
                {
                    foreach (var item in source)
                    {
                        var status = Next(item);
                        if (status.IsStopped())
                            return status;
                    }
                    return ChainStatus.Flow;
                }
            }

            ChainStatus Optimizations.IHeadStart<T>.Execute<Enumerable, Enumerator>(Enumerable source)
            {
                foreach (var input in source)
                {
                    checked
                    {
                        index++;
                    }

                    if (index >= count)
                        return ChainStatus.Stop | Next(input);

                    var status = Next(input);
                    if (status.IsStopped())
                        return status;
                }
                return ChainStatus.Flow;
            }

        }
    }
}

File: src/Cistern.Linq.FSharp/Links/take.fs
namespace Cistern.Linq.FSharp.Links

open Cistern.Linq

type TakeActivity<'T>(count:int, next) =
    inherit Cistern.Linq.Activity<'T,'T>(next)

    let mutable index = 0

    override __.ProcessNext input =
        if index >= count then
            ChainStatus.Stop
        else
            index <- Checked.(+) index 1
            if index >= count then
                ChainStatus.Stop ||| base.Next input
            else
                base.Next input

    override __.ChainComplete status =
        if not (status.HasFlag ChainStatus.Stop) && index < count then
            raise (System.InvalidOperationException (sprintf "tried to take The input sequence has an insufficient number of elements. %d past the end of the seq" (count-index)))

        base.ChainComplete status

type Take<'T>(count:int) =
    interface Cistern.Linq.ILink<'T, 'T> with
        member __.Compose activity = 
            upcast TakeActivity(count, activity)
